### PR TITLE
Task-41416_4117: Implement MalwareDetection feature

### PR DESCRIPTION
**Pull request by azayati**
This PR:
- Implements Malware detection rest service which allows to process "files" or "jcr nodes" detected as malware ,the process is right now only sending a notification to the last updater of the infected file. 
- Implements FileMalwareDetectionServiceConnector for the "files" detection case, for example after uploading a wiki attachment, a FS file example is generated into /gatein/data/files/2/2/d/b/1/2/3/3/22db1233ceb84833ac44cc36f630fb1c of the server,  the last part of the file path is the file checksum, the connector will detect this checksum, get the corresponding files using FileService then send the notification.

### CODE REVIEW

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    String fileChecksum = getInfectedFileChecksum(infectedFilePath);
+    try {
+      List<FileItem> infectedFileItems = fileService.getFilesByChecksum(fileChecksum);
+      for (FileItem infectedFileItem : infectedFileItems) {
+        if (infectedFileItem != null && infectedFileItem.getFileInfo() != null && infectedFileItem.getFileInfo().getId() != null) {
+          fileService.deleteFile(infectedFileItem.getFileInfo().getId());
+        }
+      }
+    } catch (Exception e) {
+      LOGGER.error("Error when trying to clean the infected item", e);
```
Comment (plamarque) [inline component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java]: shouldn't we log some identifier of the infected item to help troubleshoot ?

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      for (FileItem infectedFileItem : infectedFileItems) {
+        if (infectedFileItem != null && infectedFileItem.getFileInfo() != null) {
+          Map<String, String> infectedItem = new HashMap<String, String>();
+          infectedItem.put(INFECTED_ITEM_NAME, infectedFileItem.getFileInfo().getName());
+          infectedItem.put(INFECTED_ITEM_LAST_MODIFIER, infectedFileItem.getFileInfo().getUpdater());
+          infectedItems.add(infectedItem);
+        }
+      }
+    } catch (Exception e) {
+      LOGGER.error("Error when trying to get the infected item informations", e);
```
Comment (plamarque) [inline component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java]: shouldn't we log some identifier of the infected item to help troubleshoot ?

File: extension/notification/src/main/resources/locale/notification/template/Notification_en.properties [hunk 3]
```diff
@@ -278,3 +278,15 @@ UINotification.title.DlpUserRestoredItemPlugin=Item has been validated
...
 Notification.subject.DlpUserRestoredItemPlugin=Item has been validated
 Notification.intranet.message.DlpUserRestoredItemPlugin=The item named <strong>{0}</strong> has been analyzed during a quarantine phase. It has been validated and restored in his initial location.
+
+
+#############################################################################
+#                         MalwareDetectionPlugin                         #
+#############################################################################
+UINotification.label.group.MalwareDetection=Malware detection
+UINotification.label.MalwareDetectionPlugin=Malware detection
+UINotification.title.MalwareDetectionPlugin=Item is infected
```
Comment (plamarque) [inline extension/notification/src/main/resources/locale/notification/template/Notification_en.properties]: should we simply replace "item" by "file" or "document" ? seems more user friendly

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java (hunk 2 above)
Comment (azayati) [reply component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java]: yes I will rework the logs with the good format, it is just the draft :)

File: extension/notification/src/main/resources/locale/notification/template/Notification_en.properties (hunk 3 above)
Comment (azayati) [reply extension/notification/src/main/resources/locale/notification/template/Notification_en.properties]: I have used the same label of DLP, I will see with PO to put the good label

Comment (azayati) [reply extension/notification/src/main/resources/locale/notification/template/Notification_en.properties]: Modified @plamarque

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java (hunk 2 above)
Comment (azayati) [reply component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java]: I have added one log [here](https://github.com/Meeds-io/social/pull/618/files#diff-8c188b1765721990c9d74eed2edd9399e5a54bc0e38a7488a31da8df96fcf6beR38) when detecting a file as a malware

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java (hunk 1 above)
Comment (rdenarie) [reply component/service/src/main/java/org/exoplatform/social/service/malwareDetection/connector/FileMalwareDetectionServiceConnector.java]: :+1:

File: component/service/src/test/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  
+  private MalwareDetectionService malwareDetectionService;
+
+  public void setUp() throws Exception {
+    super.setUp();
+    malwareDetectionService = getContainer().getComponentInstanceOfType(MalwareDetectionService.class);
+  }
+
+  public void testProcessInfectedItem() {
+    String infectedItemPath = "/gatein/data/files/d/e/4/c/e/2/d/7/dedeerere673r777";
```
Comment (rdenarie) [inline component/service/src/test/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionServiceTest.java:18]: Can you make a test with more than one infected item

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: component/service/src/test/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionServiceTest.java (hunk 4 above)
Comment (azayati) [reply component/service/src/test/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionServiceTest.java:18]: Right now the test done is for the MalwareDetectionService which process only one infected item. We can test MalwareDetectionServiceRestService also in order to test the case of a list of infected items

**Review by rdenarie — APPROVED** [review]

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionService.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    else if (infectedItemPath.contains(getJcrNodePathSeparator(infectedItemPath) + JCR_CONNECTOR + getJcrNodePathSeparator(infectedItemPath) + VALUES + getJcrNodePathSeparator(infectedItemPath))) {
+      MalwareDetectionServiceConnector jcrMalwareDetectionServiceConnector = (MalwareDetectionServiceConnector) getConnectors().get(JCR_CONNECTOR);
+      if(jcrMalwareDetectionServiceConnector != null && jcrMalwareDetectionServiceConnector.isEnable()) {
+        jcrMalwareDetectionServiceConnector.sendInfectedItemNotification(infectedItemPath);
+        jcrMalwareDetectionServiceConnector.cleanInfectedItem(infectedItemPath);
+      }
+    }
+  }
+  
+  public static String getJcrNodePathSeparator(String infectedFilePath) {
```
Comment (rdenarie) [inline component/service/src/main/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionService.java]:
```suggestion
  public static String getPathSeparator(String infectedFilePath) {
```
As this is not related to JCR we should not use 'JCR' in method name.

File: component/service/src/main/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionService.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        fileMalwareDetectionServiceConnector.cleanInfectedItem(infectedItemPath);
+      }
+    }
+    else if (infectedItemPath.contains(getJcrNodePathSeparator(infectedItemPath) + JCR_CONNECTOR + getJcrNodePathSeparator(infectedItemPath) + VALUES + getJcrNodePathSeparator(infectedItemPath))) {
+      MalwareDetectionServiceConnector jcrMalwareDetectionServiceConnector = (MalwareDetectionServiceConnector) getConnectors().get(JCR_CONNECTOR);
+      if(jcrMalwareDetectionServiceConnector != null && jcrMalwareDetectionServiceConnector.isEnable()) {
+        jcrMalwareDetectionServiceConnector.sendInfectedItemNotification(infectedItemPath);
+        jcrMalwareDetectionServiceConnector.cleanInfectedItem(infectedItemPath);
+      }
+    }
```
Comment (rdenarie) [inline component/service/src/main/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionService.java]: getPathSeparator should be called once at the beginning of the function, and stored in a local variable, so that it will prevent to call a lot of contains on String

Comment (rdenarie) [reply component/service/src/main/java/org/exoplatform/social/service/malwareDetection/MalwareDetectionService.java]: Second question : 
I think it should be preferable to do something like this : 

for connector : connectors {
   if (connector.isConcernedBy(infectedFilePath)) {
    do stuff
  }
}

So that, you have not the logic in the master MalwareDetectionService and each connector can say if he have authority to clean the infected file.

My idea is to remove all information about JCR from social module

**Review by rdenarie — CHANGES_REQUESTED** [review]

**Review by rdenarie — APPROVED** [review]
